File: src/commands/tensorboard.js
const { spawn } = require('child_process')

const WML = require('./../api/wml')
const loadCredentials = require('./../utils/loadCredentials')
const optionsParse = require('./../utils/optionsParse')
const cosEndpointBuilder = require('./../utils/cosEndpointBuilder')

module.exports = async options => {
  const parser = optionsParse()
  parser.add('model_id')
  parser.add(['--config', '-c'])
  parser.add([true, 'help', '--help', '-help', '-h'])
  const ops = parser.parse(options)

  if (ops.help) {
    console.log('cacli tensorboard <model_id>')
    return process.exit()
  }

  if (!ops.model_id) {
    console.log('No Model ID provided')
    console.log('Usage: cacli tensorboard <model_id>')
    return process.exit(1)
  }

  const config = await loadCredentials()

  const wml = new WML(config)

  const run = await wml.getTrainingRun(ops.model_id)

  const {
    bucket,
    model_location
  } = run.entity.training_results_reference.location

  const { region, access_key_id, secret_access_key } = config.credentials.cos

  const cosConfig = {
    endpoint: cosEndpointBuilder(region, true),
    accessKeyId: access_key_id,
    secretAccessKey: secret_access_key
  }

  const tensorboard = spawn(
    'tensorboard',
    [`--logdir=s3://${bucket}/${model_location}`],
    {
      env: {
        ...process.env,
        AWS_ACCESS_KEY_ID: cosConfig.accessKeyId,
        AWS_SECRET_ACCESS_KEY: cosConfig.secretAccessKey,
        S3_ENDPOINT: cosConfig.endpoint.replace(/^(https:\/\/)/, '')
      }
    }
  )

  tensorboard.stderr.on('data', data => {
    console.log(data.toString())
  })
}

File: cacli/cmd/tensorboard/tensorboard.go
package tensorboard

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"syscall"
	"time"

	"github.com/briandowns/spinner"
	"github.com/cloud-annotations/training/cacli/cmd/login"
	"github.com/cloud-annotations/training/cacli/e"
	"github.com/spf13/cobra"
)

var s = spinner.New(spinner.CharSets[14], 60*time.Millisecond)

func Run(cmd *cobra.Command, args []string) {
	if len(args) < 1 {
		e.Exit(errors.New("no `Model ID` provided\nUsage: cacli tensorboard <model_id>"))
	}

	modelID := args[0]

	session := login.AssertLoggedIn()

	s.Suffix = " Getting TensorBoard information..."
	s.Start()

	model, err := session.GetTrainingRun(modelID)
	if err != nil {
		e.Exit(err)
	}

	bucket := model.Entity.TrainingResultsReference.Location.Bucket
	modelLocation := model.Entity.TrainingResultsReference.Location.ModelLocation

	s3endpoint, err := session.GetEndpointForBucket(bucket)
	if err != nil {
		e.Exit(err)
	}
	s3endpoint = strings.ReplaceAll(s3endpoint, "https://", "")

	binary, err := exec.LookPath("tensorboard")
	if err != nil {
		e.Exit(errors.New("tensorboard not installed"))
	}

	logdir := fmt.Sprintf("s3://%s/%s", bucket, modelLocation)
	arguments := []string{"tensorboard", "--logdir", logdir}
	env := os.Environ()
	env = append(env, fmt.Sprintf("AWS_ACCESS_KEY_ID=%s", session.AccessKeyID))
	env = append(env, fmt.Sprintf("AWS_SECRET_ACCESS_KEY=%s", session.SecretAccessKey))
	env = append(env, fmt.Sprintf("S3_ENDPOINT=%s", s3endpoint))
	env = append(env, "TF_CPP_MIN_LOG_LEVEL=3")

	syscall.Exec(binary, arguments, env)
}
